src/spikeinterface/core/analyzer_extension_core.py

```diff
@@ -224,18 +237,75 @@ def _select_extension_data(self, unit_ids):
 
         return new_data
 
-    def get_waveforms_one_unit(
-        self,
-        unit_id,
-        force_dense: bool = False,
+    def _merge_extension_data(
+        self, units_to_merge, new_unit_ids, new_sorting_analyzer, keep_mask=None, verbose=False, **job_kwargs
     ):
+        new_data = dict()
+
+        waveforms = self.data["waveforms"]
+        some_spikes = self.sorting_analyzer.get_extension("random_spikes").get_random_spikes()
+
+        if keep_mask is not None:
+            spike_indices = self.sorting_analyzer.get_extension("random_spikes").get_data()
+            valid = keep_mask[spike_indices]
+            some_spikes = some_spikes[valid]
+            waveforms = waveforms[valid]
+
+        sparsity = new_sorting_analyzer.sparsity
+
+        if sparsity is not None:
+
+            max_num_chans = np.sum(new_sorting_analyzer.sparsity.mask, 1).max()
+            new_data["waveforms"] = waveforms.copy()
+            for to_be_merged, unit_id in zip(units_to_merge, new_unit_ids):
+                new_channel_ids = sparsity.unit_id_to_channel_ids[unit_id]
+                new_waveforms, spike_indices = self.get_some_waveforms(
+                    new_channel_ids, to_be_merged, kept_waveforms=new_data["waveforms"], kept_spikes=some_spikes
+                )
+                num_chans = new_waveforms.shape[2]
+                new_data["waveforms"][spike_indices, :, :num_chans] = new_waveforms
+            new_data["waveforms"] = new_data["waveforms"][:, :, :max_num_chans]
+        else:
+            new_data["waveforms"] = waveforms
+
+        return new_data
+
+    def get_waveforms_one_unit(self, unit_id, force_dense: bool = False, kept_waveforms=None, kept_spikes=None):
```

Comment:
I know, but what is your suggestion ? We need to provide to these functions the subarrays of valid waveforms and spikes, in order to avoid constently recomputing them by slicing. This is much faster.